### PR TITLE
Re-arranges fetcher locks -> per peer level

### LINKED ISSUE #5902: Blocks fetcher - re-arrange locks
Per @prestonvanloon's suggestion:
> This is a global lock that can be blocked by a particular peer's rate limit. In other words, if we are rate limited by one peer, we are blocking all outgoing requests until that peer's rate limit is satisfied.
https://github.com/prysmaticlabs/prysm/blob/e16f3842867e0c73e48fe5f0bd12b22e9c7b1642/beacon-chain/sync/initial-sync/blocks_fetcher.go#L372
Do you want to rearrange this global lock to avoid this type of behavior?

### PR DESCRIPTION
**Pull request by farazdagi**
**What type of PR is this?**

Bug fix

**What does this PR do? Why is it needed?**
- locking on fetch requests was too restrictive: instead of locking rate limiter read/write on per peer level, it used a single lock, thus being too aggresive.

**Which issues(s) does this PR fix?**

Fixes #5902 

**Other notes for review**
- Implementation relies on map of locks - each peer gets its own lock, so fetcher can multiplex access among peers easily
- Additional routine to remove stale locks (for peers that once appeared in fetcher but then disappeared and their locks haven't been used for a long period of time) is added.

### CODE REVIEW

File: beacon-chain/sync/initial-sync/blocks_fetcher.go [hunk 1]
```diff
@@ -137,6 +149,21 @@ func (f *blocksFetcher) loop() {
...
 
+	// Periodically remove stale peer locks.
+	go func() {
+		ticker := time.NewTicker(peerLocksPollingInterval)
+		for {
+			select {
+			case <-ticker.C:
+				f.removeStalePeerLocks(peerLockMaxAge)
+			case <-f.ctx.Done():
+				ticker.Stop()
```
Comment (terencechain) [inline beacon-chain/sync/initial-sync/blocks_fetcher.go:160]: Any reason to not return `ctx.Err()` here?

Comment (farazdagi) [reply beacon-chain/sync/initial-sync/blocks_fetcher.go:160]: My purpose of listening to `ctx.Done()` was to make sure we do not leak goroutines when fetcher is done. So, that return will simply wrap up surrounding goroutine and not leave it hanging - so, I am not sure what returning `ctx.Err()` will achieve here. Can you please elaborate more what should be done?

File: beacon-chain/sync/initial-sync/blocks_fetcher.go [hunk 2]
```diff
@@ -437,6 +475,35 @@ func (f *blocksFetcher) requestBlocks(
...
+}
+
+// removeStalePeerLocks is a cleanup procedure which removes stale locks.
+func (f *blocksFetcher) removeStalePeerLocks(age time.Duration) {
+	f.Lock()
+	defer f.Unlock()
+	for peerID, lock := range f.peerLocks {
+		if time.Since(lock.accessed) >= age {
+			lock.Lock()
+			f.peerLocks[peerID] = nil
```
Comment (nisdas) [inline beacon-chain/sync/initial-sync/blocks_fetcher.go:500]: if you are going to delete it anyway is there any point of setting this to nil ?

Comment (farazdagi) [reply beacon-chain/sync/initial-sync/blocks_fetcher.go]: Nice catch - I guess it was "garbage collector worshiping gone wrong", updated!

File: beacon-chain/sync/initial-sync/blocks_fetcher.go (hunk 1 above)
Comment (terencechain) [reply beacon-chain/sync/initial-sync/blocks_fetcher.go:160]: Got it. It seems like the error case is not a valid concern here. Sounds good to me then

**Pull request merged**